pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,281 @@
+		return nil, fmt.Errorf("listing containers: %w", err)
+	}
+
+	ret := make([]*runtimeclient.ContainerData, 0, len(containers))
+	for _, container := range containers {
+		if isSandboxContainer(container) {
+			log.Debugf("%scontainer %q is a sandbox container. Temporary skipping it", logPrefix, container.ID())
+			continue
+		}
+
+		task, err := getContainerTask(container)
+		if err != nil {
+			log.Debugf("%sgetting containerTask for container %q: %s", logPrefix, container.ID(), err)
+			continue
+		}
+
+		containerData, err := taskAndContainerToContainerData(task, container)
+		if err != nil {
+			log.Debugf("%screating containerData for container %q: %s", logPrefix, container.ID(), err)
+			continue
+		}
+
+		ret = append(ret, containerData)
+	}
+
+	return ret, nil
+}
+
+func (c *ContainerdClient) GetContainer(containerID string) (*runtimeclient.ContainerData, error) {
+	container, err := c.getContainer(containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("listing labels of container %q: %w", container.ID(), err)
+	}
+
+	// State is getting set to `Created` here for the following reasons:
+	// 1. GetContainer is only getting called on new created containers
+	// 2. We would need to get the Task for the Container. containerd needs to aquire a mutex
+	//    that is currently hold by the creating process, which we interrupted -> deadlock
+	containerData := &runtimeclient.ContainerData{
+		ID:      container.ID(),
+		Name:    getContainerName(container),
+		State:   runtimeclient.StateCreated,
+		Runtime: runtimeclient.ContainerdName,
+	}
+	runtimeclient.EnrichWithK8sMetadata(containerData, labels)
+	return containerData, nil
+}
+
+func (c *ContainerdClient) GetContainerDetails(containerID string) (*runtimeclient.ContainerDetailsData, error) {
+	containerID, err := runtimeclient.ParseContainerID(runtimeclient.ContainerdName, containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	containerData, container, task, err := c.getContainerDataAndContainerAndTask(containerID)
+	if err != nil {
+		return nil, err
+	}
+	if task.pid == 0 {
+		return nil, fmt.Errorf("got zero pid")
+	}
+
+	spec, err := container.Spec(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("getting spec for container %q: %w", containerID, err)
+	}
+
+	mountData := make([]runtimeclient.ContainerMountData, len(spec.Mounts))
+	for i := range spec.Mounts {
+		mount := spec.Mounts[i]
+		mountData[i] = runtimeclient.ContainerMountData{
+			Source:      mount.Source,
+			Destination: mount.Destination,
+		}
+	}
+
+	return &runtimeclient.ContainerDetailsData{
+		ContainerData: *containerData,
+		Pid:           int(task.pid),
+		CgroupsPath:   spec.Linux.CgroupsPath,
+		Mounts:        mountData,
+	}, nil
+}
+
+func (c *ContainerdClient) getContainerDataAndContainerAndTask(containerID string) (*runtimeclient.ContainerData, containerd.Container, *containerTask, error) {
```

Comment:
nit: If we aren't planning to reuse it I feel this is too specific and we should move all of this in `GetContainerDetails` directly. For me that would be more readable!

Reply:
I created the method not only for reusability (which we don't reuse anymore :D).
The reason was also to separate the code block from the function that calls it. For readability and keeping the function body small